collections/_article/rendering-acyclic-graph.md

```diff
@@ -0,0 +1,279 @@
+
+#### Command Buffers
+
+Recording work for the GPU and submitting it for execution is a very explicit operation in Vulkan. [Command buffers](https://docs.vulkan.org/spec/latest/chapters/cmdbuffers.html) (or command lists in D3D12) are the objects where all the recorded work is stored that that will executed at a later point on the GPU. These buffers can grow as much as the user desires and they won't be run until they're submitted for execution to a command queue. This implies it's essentially possible to record from multiple sources, even multiple threads, and submit the work to the GPU once it's ready with proper synchronization.
+
+#### Render Passes
+
+Drawing something with Vulkan requires a lot of information upfront in the form of a "render pass": an object that contains references to the textures that the GPU will use as targets, their initial and final states (texture memory layouts) and much more. This is a stark contrast from older APIs like OpenGL or even D3D12, where the render target can be changed during command recording and does not require the creation of an object ahead of time. While a render pass is active, there's also various restrictions on what operations can be recorded on the command buffer. The render pass must end before being allowed to use all types of operations again.
+
+#### Barriers
+
+The biggest change compared to previous APIs is that synchronization between commands is no longer automatically deduced by the driver and it must be manually implemented by the programmers using Vulkan. If you want to read more about this topic in detail, I highly recommend Hans-Kristian's [blog post about Vulkan Synchronization](https://themaister.net/blog/2019/08/14/yet-another-blog-explaining-vulkan-synchronization/) and reading the [official Vulkan documentation](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCmdPipelineBarrier.html). To keep the article short, a very basic explanation is provided below.
+
+The order of execution of the recorded commands inside a command buffer **is NOT guaranteed to complete in the order they were submitted in**: the GPU can reorder these commands in whatever order it thinks is best to complete the job as quickly as possible. To compensate for this, the programmer must manually insert **[synchronization barriers](https://gpuopen.com/learn/vulkan-barriers-explained/)** in the command buffer that allow specifying in detail which commands should be completed or start by specifying the scope **before** and **after**. The scope includes multiple concepts such as the execution stage (e.g. drawing, compute, transfer), the type of access (e.g. read or write) and the affected memory regions (globally, buffers or textures). On top of that, barriers are capable of transitioning textures from one memory layout to another, which is a requirement to be able to use textures in different commands optimally.
```

Comment:
```suggestion
The order of execution of the recorded commands inside a command buffer **is NOT guaranteed to complete in the order they were submitted**: the GPU can reorder these commands in whatever order it thinks is best to complete the job as quickly as possible. To compensate for this, the programmer must manually insert **[synchronization barriers](https://gpuopen.com/learn/vulkan-barriers-explained/)** in the command buffer that allow specifying in detail which commands should be completed or start by specifying the scope **before** and **after**. The scope includes multiple concepts such as the execution stage (e.g. drawing, compute, transfer), the type of access (e.g. read or write) and the affected memory regions (globally, buffers or textures). On top of that, barriers are capable of transitioning textures from one memory layout to another, which is a requirement to be able to use textures in different commands optimally.
```

Reply:
I'm not sure about "which commands should be completed or start" - should it be "which commands should be completed or started"?

Reply:
I actually wrote what you said at first and then making it go through google translate gave me this sentence instead. But what you said sounds more correct to me as well.